VMEncryption/main/DiskUtil.py

```diff
@@ -397,6 +464,47 @@ def luks_get_uuid(self, header_or_dev_path):
                 return splits[1]
         return None
 
+    def get_token_id(self,header_or_dev_path,token_name):
+        '''if LUKS2 header has token name return the id else return none.'''
+        if not header_or_dev_path or not os.path.exists(header_or_dev_path) or not token_name:
+            self.logger.log("get_token_id: invalid input, header_or_dev_path:{0} token_name:{1}".format(header_or_dev_path,token_name))
+            return None
+        luks_dump_out = self._luks_get_header_dump(header_or_dev_path)
+        tokens = self._extract_luksv2_token(luks_dump_out)
+        for token in tokens:
+            if len(token) == 2 and token[1] == token_name:
+                return token[0]
+        return None
+
+    def restore_luks2_token(self, device_name=None):
+        '''this function restoring token type Azure_Disk_Encryption_BackUp to Azure_Disk_Encryption, 
+        this function acts on 4 secenarios. [token id Azure_Disk_Encryption, token id Azure_Disk_Encryption_BackUp, resote_action, Scenario]'''
+        '''[Y,Y,remove token type Azure_Disk_Encryption_BackUp,VM restart/failure]
+           [Y,N,do nothing,Success scenario]
+           [N,N,do nothing,No token present]
+           [N,Y,move token type Azure_Disk_Encryption_BackUp to Azure_Disk_Encryption,VM restart/failure]'''
+        if not device_name or not os.path.exists(self.get_device_path(device_name)):
+            self.logger.log(level=CommonVariables.WarningLevel,msg="restore_luks2_token invalid input. device_name = {0}".format(device_name))
+            return
+        device_path = self.get_device_path(device_name)
+        ade_token_id = self.get_token_id(header_or_dev_path=device_path,token_name=CommonVariables.AzureDiskEncryptionToken)
+        ade_token_id_backup = self.get_token_id(header_or_dev_path=device_path,token_name=CommonVariables.AzureDiskEncryptionBackUpToken)
+        if not ade_token_id_backup:
+            #do nothing
+            return
+        if ade_token_id:
+            #remove backup token id
+            self.remove_token(device_name=device_name,token_id=ade_token_id_backup)
+            return
+        self.logger.log("resotre luks2 token for device {0} is started.".format(device_name))
+        #read from backup and update AzureDiskEncryptionToken
+        data = self.read_token(device_name=device_name,token_id=ade_token_id_backup)
+        data['type']=CommonVariables.AzureDiskEncryptionBackUpToken
```

Comment:
they token type should not be ...BackupToken, because it is written to id 5 in next line.

Reply:
done